test/gtest/ucx_tests.cpp

```diff
@@ -49,168 +44,286 @@ namespace nixl {
-        EXPECT_EQ(NIXL_SUCCESS, rAgent->getLocalMD(rMeta));
-
-        std::string sMeta_remote;
-        EXPECT_EQ(NIXL_SUCCESS, sAgent->loadRemoteMD(rMeta, sMeta_remote));
-        std::string rMeta_remote;
-        EXPECT_EQ(NIXL_SUCCESS, rAgent->loadRemoteMD(sMeta, rMeta_remote));
-
-        nixlBasicDesc sReq_src;
-        nixl_xfer_dlist_t sReq_descs(DRAM_SEG);
-        nixl::fillRegList(sReq_descs, sReq_src, sData);
-
-        if (receiver_failure) {
-            rAgent->deregisterMem(rDlist, &rExtraParams);
-            rAgent.reset();
-        }
+nixl_status_t
+UcxTestFixture::Agent::wait_for_completion(nixlXferReqH* req_handle) {
+    nixl_status_t status;
 
-        nixlBasicDesc rReq_dst;
-        nixl_xfer_dlist_t rReq_descs(DRAM_SEG);
-        nixl::fillRegList(rReq_descs, rReq_dst, rData);
-
-        sExtraParams.notifMsg = "notification";
-        sExtraParams.hasNotif = true;
-        nixlXferReqH* req_handle;
-        EXPECT_EQ(NIXL_SUCCESS,
-                  sAgent->createXferReq(NIXL_WRITE, sReq_descs, rReq_descs,
-                                        "receiver", req_handle, &sExtraParams));
-        if (receiver_failure) {
-            // the error may be returned immediately or later
-            nixl_status_t status = sAgent->postXferReq(req_handle);
-            EXPECT_TRUE((status == NIXL_ERR_REMOTE_DISCONNECT) ||
-                        (status == NIXL_IN_PROG));
-        } else {
-            EXPECT_LE(0, sAgent->postXferReq(req_handle));
-        }
+    do {
+        status = m_priv->getXferStatus(req_handle);
+    } while (status == NIXL_IN_PROG);
+
+    m_priv->releaseXferReq(req_handle);
+    return status;
+}
+
+nixl_status_t
+UcxTestFixture::Agent::wait_for_notif(const std::string& expectedNotif) {
+    nixl_notifs_t notif_map;
+
+    do {
+        EXPECT_EQ(NIXL_SUCCESS, m_priv->getNotifs(notif_map));
+    } while (notif_map.empty());
+
+    std::vector<std::string> notifs = notif_map[m_MetaRemote];
+    EXPECT_EQ(1,             notifs.size());
+    EXPECT_EQ(expectedNotif, notifs.front());
+    return NIXL_SUCCESS;
+}
+
+void UcxTestFixture::Agent::fillData() {
+    m_mem.fillData();
+}
+
+bool UcxTestFixture::Agent::dataCmp(const UcxTestFixture::Agent& other) const {
+    return m_mem.m_data == other.m_mem.m_data;
+}
+
+UcxTestFixture::UcxTestFixture() {
+    // Set up test environment
+    m_plugin_dir_backup = getenv("NIXL_PLUGIN_DIR");
+
+    // Load plugins from build directory
+    std::string plugin_dir = std::string(BUILD_DIR) + "/src/plugins/ucx";
+    setenv("NIXL_PLUGIN_DIR", plugin_dir.c_str(), 1);
+
+    std::cout << "set NIXL_PLUGIN_DIR: " << getenv("NIXL_PLUGIN_DIR")
+              << std::endl;
+}
+
+UcxTestFixture::~UcxTestFixture() {
+    setenv("NIXL_PLUGIN_DIR", m_plugin_dir_backup.c_str(), 1);
+    std::cout << "restore NIXL_PLUGIN_DIR: " << getenv("NIXL_PLUGIN_DIR")
+              << std::endl;
+}
+
+template<UcxTestFixture::TestType test_type, enum nixl_xfer_op_t op>
```

Comment:
Out of curiosity, what is the advantage of using the template vs. just passing these as arguments here?

Reply:
from functionality perspective - no difference, the main advantage - compile time optimizations, in this particular case helper functions which just convert one constant to other one will be completely compiled out.